wolfcrypt/src/rsa.c

```diff
@@ -1000,374 +1034,472 @@ static int wc_RsaFunction(const byte* in, word32 inLen, byte* out,
+                return NitroxRsaSSL_Verify(in, inLen, out, outLen, key);
+            }
+        }
+    #endif
 
+        key->state = RSA_STATE_DECRYPT_UNPAD;
 
-int wc_RsaPrivateDecrypt(const byte* in, word32 inLen, byte* out, word32 outLen,
-                     RsaKey* key)
-{
-    int plainLen;
-    byte*  tmp;
-    byte*  pad = 0;
+        /* verify the tmp ptr is NULL, otherwise indicates bad state */
+        if (key->tmp != NULL) {
+            ERROR_OUT(BAD_STATE_E);
+        }
 
-#ifdef HAVE_CAVIUM
-    if (key->magic == WOLFSSL_RSA_CAVIUM_MAGIC)
-        return CaviumRsaPrivateDecrypt(in, inLen, out, outLen, key);
-#endif
+        /* if not doing this inline then allocate a buffer for it */
+        key->tmpLen = inLen;
+        if (outPtr == NULL) {
+            key->tmp = (byte*)XMALLOC(inLen, key->heap, DYNAMIC_TYPE_RSA);
+            if (key->tmp == NULL) {
+                ERROR_OUT(MEMORY_E);
+            }
+            XMEMCPY(key->tmp, in, inLen);
+        }
+        else {
+            key->tmp = out;
+        }
+        ret = wc_RsaFunction(key->tmp, inLen, key->tmp, &key->tmpLen,
+                                                        rsa_type, key, rng);
+        if (ret < 0) {
+            break;
+        }
+        /* fall through */
+    case RSA_STATE_DECRYPT_UNPAD:
+    {
+        byte* pad = NULL;
+        key->state = RSA_STATE_DECRYPT_RES;
+        ret = wc_RsaUnPad_ex(key->tmp, key->tmpLen, &pad, pad_value, pad_type,
+                                        hash, mgf, label, labelSz, key->heap);
+        if (ret > 0 && ret <= (int)outLen && pad != NULL) {
+            /* only copy output if not inline */
+            if (outPtr == NULL) {
+                XMEMCPY(out, pad, ret);
+            }
+            else {
+                *outPtr = pad;
+            }
+        }
+        else if (ret >= 0) {
+            ret = RSA_BUFFER_E;
+        }
+        if (ret < 0) {
+            break;
+        }
+        /* fall through */
+    }
+    case RSA_STATE_DECRYPT_RES:
+        key->state = RSA_STATE_NONE;
+    #if defined(WOLFSSL_ASYNC_CRYPT) && defined(HAVE_CAVIUM)
+        if (key->asyncDev.marker == WOLFSSL_ASYNC_MARKER_RSA) {
+            ret = key->tmpLen;
+        }
+    #endif
+        break;
+    default:
+        ret = BAD_STATE_E;
+    }
 
-    tmp = (byte*)XMALLOC(inLen, key->heap, DYNAMIC_TYPE_RSA);
-    if (tmp == NULL) {
-        return MEMORY_E;
+    /* if async pending then return and skip done cleanup below */
+    if (ret == WC_PENDING_E) {
+        return ret;
     }
 
-    XMEMCPY(tmp, in, inLen);
+done:
 
-    if ( (plainLen = wc_RsaPrivateDecryptInline(tmp, inLen, &pad, key) ) < 0) {
-        XFREE(tmp, key->heap, DYNAMIC_TYPE_RSA);
-        return plainLen;
+    key->state = RSA_STATE_NONE;
+    if (key->tmp) {
+        /* if not inline */
+        if (outPtr == NULL) {
+            ForceZero(key->tmp, key->tmpLen);
+            XFREE(key->tmp, key->heap, DYNAMIC_TYPE_RSA);
```

Comment:
Shouldn't wc_FreeRsaKey() also release key->tmp if it's still valid?  In PENDING_E state if the user never calls back in to complete key->tmp will leak if FreeRsaKey() doesn't also release.